## 异常

###### 1.异常：

对问题的描述；将问题进行对象的封装。   

异常分两种：（1）编译时被检测的异常（该异常在编译时，在抛出非RuntimeException时，没有声明，没有处理，则编译失败)  
(2)在运行时异常（RuntimeException以及去子类）

###### 2. 异常的特点：
具有可抛性
###### 
3. throw和throws的区别：
throws：使用在函数上；后面跟异常类（可以多个）。
throw：使用在函数内，后面跟异常对象。   
当函数内容有throw抛出异常对象，并未进行try处理，必须要在函数上声明，否则编译失败，（RuntimeException除外）。

###### 4.异常处理：

格式：（1）try{需要被检测部分}catch{处理方式}finally{一定执行部分}
（2）try{}catch{}（3）try{}finall{}  
 
对多异常的处理：
（1）声明异常时，建议声明更为具体的异常，这样处理得更具体
（2）对方声明几个异常，就应该有几个catch块，如果多个catch块中的异常出现继承关系，父类异常catch块放在最下面。

###### 5.子类在覆盖中的体现：

（1）子类在覆盖父类时，如果父类的方法抛出异常，那么子类的覆盖方法，只能抛出父类的异常（不能抛出父类没有的异常）；子类中发生新异常就必须在内部处理，不能抛出（只能用try）。

###### 6.异常处理规则：

（1）处理方式：try或throws    

（2）当捕获到的异常，在本功能处理不了，可以继续在catch抛出；


```

try

{
 
   throw new Exception();

}

catch(Exception e)

{

    throw e;//要有声明，或内部处理。

}

/*当在函数内部出现了throw抛出异常对象，
 
那么就必须给对应的处理动作
 （try catch\在函数上声明让调用者处理）
*/

```


如果该异常处理不了，但并不属于该功能出现的异常，可以将异常转换后，再抛出与该功能相关的异常


```

try

{

    throw new AException();

}

catch(Aexception e)

{
    //先对AException进行处理。

    throw new BException();

}

```
